---
title: "DATABASE FINAL PROJECT"
author: "Sowmya, Aditya"
date: "12/17/2021"
output: html_document
---

```{R setup}
library(odbc)
library(skimr)
library(ggplot2)
```


Setting connection to our database in Azure using an R library called DBI. 
```{R dbi con}

library(DBI)
con <- dbConnect(odbc(), Driver = "SQL Server", Server = "gene.database.windows.net", 
    Database = "GENE_DB", UID = "sowmyap", PWD = "HA78BGb&s4%m", 
    Port = 1433)

```

Query1: Find the total gene counts for each spot. 
```{sql, connection = con, output.var = 'spotcountsdf'}

/*spotcountsdf is the R object that is outputted from this query */
SELECT SUM(counts) AS spot_counts, spotid FROM gene_counts
GROUP BY spotid;
```

The skim function in R from the skimr library gives a comprehensive summary of 

spotcountsdf as seen below. 
```{R spotcountsdf}

skim(spotcountsdf)

```

```{sql, connection = con, output.var = 'query2'}

SELECT TOP 1 * FROM coldata;
```

```{R q2 }
query2
```

Query 2: We want to find the proportion of zeros to nonzero cell count in our db

```{sql, connection = con, output.var = 'proportion'}

SELECT zeros/nonzeros FROM Z,ZN;


```


```{R zn }
proportion
```

Query 3: Find all image pixels associated with layer 1. 
```{sql, connection = con, output.var = 'layer1' }
SELECT pxl_row_in_fullres, pxl_col_in_fullres FROM coldata
WHERE ground_truth = 'Layer1' ;
```

```{R layer 1  }
layer1
```

Now, using the ggplot library in R, we can plot these pixel values to figure out the shape of layer 1 in our tissue image as seen below. 
```{R layer 1 shape }

ggplot(data=layer1, aes(
            x=pxl_row_in_fullres, y=pxl_col_in_fullres) 
       ) + geom_point()
```

Query 4: 
Give the number of expressed genes for all spots that are located in the lower right quadrant of the spot grid. (all genes for which counts > 0).
```{sql,connection = con, output.var = 'max_array_row'  }

SELECT MAX(array_row) FROM coldata;

```

```{R max array row}
max_array_row
```

```{sql,connection = con, output.var = 'max_array_col' }

SELECT MAX(array_col) FROM coldata;
```

```{R max array col}
print(max_array_col)
```


```{sql,connection = con, output.var = 'query4' }
SELECT  DISTINCT geneid FROM coldata, gene_counts WHERE
gene_counts.spotid = coldata.barcode_id AND 
counts >0 AND 
array_row > 77/2 AND
array_col > 127/ 2;

```


```{r q4 }
# list of genes
query4
```

 Select the spot with the highest proportion of mitochondrial genes. 



```{sql,connection = con, output.var = 'query5'}

SELECT TOP 1 mt/nonmt AS proportion FROM M, N
WHERE M.spotid = N.spotid;
/*ORDER BY proportion DESC  ; */

```


```{R print q5}
query5 
```
We get zero rows with mitchondrial genes, perhaps none of them were captured for this layer of the brain. 




We have genomic database, but we also want to explore data on people who image this tissues and help create these datasets. While information on this is not readily available due to security reasons, we imagined what such relations would look like, made fake data, and have explored them here. 

For query 8, we want to explore the number of users outside of Baltimore that work with such datasets. The answer we get is 27. 


Many genes are not detected in the tissue. Let's retrieve the ones that are expressed. 

```{sql,connection = con, output.var = 'query6'}
SELECT DISTINCT gene_name FROM rowdata, gene_counts
	WHERE rowdata.gene_id = gene_counts.geneid AND gene_counts.counts >= 1


```

```{R print query6}

query6
```

```{sql,connection = con, output.var = 'query8' }


SELECT count(*) FROM users AS U ,institutions AS I, zipcode as Z
where Z.City != 'Baltimore' AND I.zipcode = Z.Zipcode AND 
u.institution_id = i.institutionid ;
```

```{R print q8}

query8
```


Query 9 gives the gene name that is expressed the most in the tissue. 
```{sql,connection = con, output.var = 'query9' }

SELECT TOP 1 SUM(counts) , gene_name FROM gene_counts , rowdata, coldata 
WHERE gene_counts.geneid = rowdata.gene_id AND gene_counts.spotid = coldata.barcode_id
GROUP BY gene_name ORDER BY SUM(counts) DESC ;

```

```{R print query9}
query9

```
For query 10, let's see who first joined the institution, and the name of the institution that has the most number of people working on spatial transcriptomics dataset.
```{sql,connection = con, output.var ='query10' }

SELECT TOP 1 fname, lname, username, institutions.name FROM Institutions, U, Users
WHERE num_users = (SELECT MAX(num_users) FROM U) AND
Institutions.institutionid = Users.institution_id AND U.institution_id = users.institution_id
ORDER BY date_joined;
```


```{R }
query10
```


Query 11 outputs the email id of all people working with a specific person stored in J. 
```{sql,connection = con, , output.var ='query11' }

SELECT users.Email_Id FROM users, J
WHERE
users.institution_id = J.institution_id;
```

```{R print q11 }
query11
```


Query 12:Print the first and last names of people who previously worked at Johns Hopkins but currently work for another institution.  

```{sql,connection = con, output.var = 'query12'}

SELECT U.fname, U.lname FROM Users AS U, Users AS V, Institutions AS I
WHERE
 U.UserID = V.UserID AND U.Institution_Id = I.institutionid AND
U.date_joined < V.date_joined AND
I.name = 'Johns Hopkins';
```

```{R print q12}
query12
```


Query 13: 
List all sample ids and brain regions where time (in hours) between scanning tissue sample and transferring data to directory was more than 48 hours, and the user it was scanned by. 

This query was interesting to figure out since it required a special function 
DATEDIFF() to calculate the differences between two dates (normal arithmetic 
subtraction doesn't work on it)
```{sql, connection = con, output.var = 'query 13'}
SELECT sample_id, br_region, scanned_by FROM TissueSample WHERE
    DATEDIFF(dy,tr_date,  scan_date) >= 2 ;
    
```

```{R print q13 }
`query 13`
```


Let's see whether we can find a certain array row in our coldata aka spot data. 
If it exists, return the number of cell counts associated with that spot. 
Let's select array row = 156 

```{sql, connection = con, output.var = 'query14'}

IF EXISTS (SELECT * FROM coldata WHERE array_row= 156)
SELECT cell_count FROM coldata WHERE array_row = 156;

```

```{R query14}
query14
```
Looks like we don't have any spots associated with array row 156. 


We want to find the platform that has on average been used in most recent times. 
Our approach to this was to parse out scan dates for each platform into
day, month and year. --> group by platform number and calculate average of day, momth and year and sort the platform numbers by these values in desc order. We finally select the TOP 1 row to find out. 
```{sql, connection = con, output.var = 'query15'}

SELECT TOP 1 AVG(CAST(MONTH(scan_date)AS int)) AS avg_month, 
AVG(CAST(YEAR(scan_date)AS int)) AS avg_year,
AVG(CAST(DAY(scan_date)AS int)) AS avg_day,
Platform_num

FROM tissuesample 
WHERE Platform_num != 'NA'
GROUP BY Platform_num
ORDER BY avg_day, avg_month, avg_year DESC;
```

```{R query15 }
query15
```


